okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.java

```diff
@@ -519,9 +528,21 @@ boolean writeOneFrame() throws IOException {
 
   void writePingFrame() {
     WebSocketWriter writer;
+    int sentPingCount;
+    int receivedPongCount;
     synchronized (this) {
       if (failed) return;
       writer = this.writer;
+      sentPingCount = this.sentPingCount;
+      receivedPongCount = this.receivedPongCount;
+      this.sentPingCount++;
+    }
+
+    if (sentPingCount > receivedPongCount) {
```

Comment:
This ping strategy is simple and understandable, but I'm concerned it could break some existing clients with already established tight ping interval.  e.g. in 3.9.1 they used 50ms within a datacentre.  This ping strategy would then turn every occasional transient network blip into a connection loss.  Likewise on a mobile device a ping interval now needs to be > 3? 5? seconds, but even then may cause losses that were otherwise recoverable.

As it's an upgrade we should probably provide guidance during the upgrade of what a minimum interval should be.

An alternative strategy we used in RSocket was to allow a configurable number of outstanding pings before disconnection, so you can keep a low ping interval to detect problems sooner, but still recover from transient blips.  However it complicates things here greatly because of the ability of servers to coalesce pongs to the final response, so you would need to use the application data in the ping/pong to send a counter or similar.

Reply:
n.b. This change is effectively changing from ∞ # of outstanding pongs to 1, so it does seem possible we will catch some clients unaware in the upgrade.

Not a blocker from me, your call.

Reply:
Yeah, this definitely needs to be considered by users as they upgrade. 

My expectation is that we'll detect a bunch of connectivity problems that were previously undetected. Hopefully we don't have false positives due to overly-aggressive ping intervals.